backend/src/external_api/cdogs/cdogs.service.ts

```diff
@@ -0,0 +1,198 @@
+        },
+      },
+    );
+
+    return response?.data?.access_token;
+  };
+
+  //--
+  //-- check the cdogs service to see if the template requred
+  //-- is cached, if it is not, upload the template and update
+  //-- the system configuration hash for the template
+  //--
+  isTemplateCached = async (apiToken: string, templateCode: string): Promise<boolean> => {
+    let result = false;
+    const config = await this.configService.findByCode(templateCode);
+
+    //-- if there's a value for the template double check
+    //-- that the the template hash is still valid
+    if (config?.configurationValue) {
+      result = await this._isCachedTemplateValid(apiToken, config.configurationValue);
+    }
+
+    return result;
+  };
+
+  //--
+  //-- upload a template to the CDOGS api, if the template exists
+  //-- the system will return the template hash, and this will
+  //-- need to be stored for the next time the document needs to be
+  //-- generated
+  //--
+  upload = async (apiToken: string, type: string, templateCode: string) => {
+    const url = `${this.baseUri}/api/v2/template`;
+
+    const template =
+      type === "HWCR"
+        ? "templates/complaint/CDOGS-HWCR-COMPLAINT-TEMPLATE-v1.docx"
+        : "templates/complaint/CDOGS-ERS-COMPLAINT-TEMPLATE-v1.docx";
+    const path = join(process.cwd(), template);
+
+    try {
+      const bodyFormData = new FormData();
+      bodyFormData.append("template", fs.createReadStream(path));
+      const headers: any = {
+        ...bodyFormData.getHeaders(),
+      };
+
+      const response: AxiosResponse = await post(apiToken, url, bodyFormData, headers);
+
+      //-- update the system with the hash provided by the CDGOS api
+      const { data: hash } = response;
+      this.configService.updateByCode(templateCode, hash);
+    } catch (error) {
+      //-- if CDOGS already has a template cached the api will throw an error
+      //-- with the existing template hash, this will need to be saved for
+      //-- the next time the document is generated
+      const {
+        response: {
+          data: { status, hash },
+        },
+      } = error;
+
+      if (status === constants.HTTP_STATUS_METHOD_NOT_ALLOWED) {
+        //-- existing template
+        this.configService.updateByCode(templateCode, hash);
+      } else {
+        //-- valid error
+        this.logger.error(`exception: unable to upload template: ${template} - error: ${error}`);
+        throw new Error(`exception: unable to upload template: ${template} - error: ${error}`);
+      }
+    }
+  };
+
+  //--
+  //-- render complaint to pdf
+  //--
+  generate = async (documentName: string, data: any, type: COMPLAINT_TYPE): Promise<AxiosResponse> => {
+    const templateCode = type === "HWCR" ? CONFIGURATION_CODES.HWCTMPLATE : CONFIGURATION_CODES.ERSTMPLATE;
+
+    try {
+      const apiToken = await this.authenticate();
+
+      if (!(await this.isTemplateCached(apiToken, templateCode))) {
```

Comment:
We shouldn't log the api token because it will be viewable to anyone scraping the logs in Kibana

Reply:
I actually just removed that unnecessary logging there because of just that :)